### ISSUE TITLE
[Draft] MultiROM support

### ISSUE THREAD

**Issue by nemanjan00** [postmarketOS/pmbootstrap#421]
**This issue is still under construction.** 

***This is draft from my memory, I need to do fact cheking.***

[MultiROM](https://github.com/Tasssadar/multirom) is dualboot script that runs after kernel and before init scripts. 

It lets users pick different ROM (that are installed by [TWRP fork](https://github.com/Tasssadar/Team-Win-Recovery-Project)). 

It is using modified [kexec-tools](https://github.com/Tasssadar/kexec-tools) for running kernel of different ROM. 

Before running another kernel, it mounts paths to proper locations, for that different ROM, 
from internal storage or USB flash (which seams like a better solution for us, right now, since we are using custom file storage)

**Comment by PabloCastellano** [comment]
Just for the record, [EFIDroid](http://efidroid.org/) is a similar project.

Some quick notes about EFIDroid:
- It is a UEFI based Android bootloader, as opposed to MultiROM which uses kexec to load a different kernel from initramfs (is this correct?)
- At the moment, only devices with Qualcomm chipsets are supported ([source](https://github.com/efidroid/projectmanagement/wiki/%5BEFIDroid%5D-Porting-a-new-device))
- Supported devices: [hammerhead](https://github.com/efidroid/projectmanagement/wiki/%5BDevice%5D-lge%7Chammerhead), [condor](https://github.com/efidroid/projectmanagement/wiki/%5BDevice%5D-motorola%7Ccondor), [thea](https://github.com/efidroid/projectmanagement/wiki/%5BDevice%5D-motorola%7Cthea), [bacon](https://github.com/efidroid/projectmanagement/wiki/%5BDevice%5D-oneplus%7Cbacon), [aries](https://github.com/efidroid/projectmanagement/wiki/%5BDevice%5D-xiaomi%7Caries)
- It uses the Android build system ([source](https://github.com/efidroid/projectmanagement/wiki/%5BEFIDroid%5D-Build-system))

**Comment by zhuowei** [comment]
@PabloCastellano The latest versions of MultiROM no longer needs the compatibility patches for the kernel being kexec'd; you only need the full-patched kernel. (see the [commit message](https://gist.github.com/Tasssadar/6766757#file-mako-aosp-implement-kexec-hardboot-patch-L412))

**Comment by deepakgangwar** [comment]
[Dual boot patcher](https://dbp.noobdev.io/) seems a good alternate option for this purpose as this does not require any major kernel modification as far as I know. The main benefits of using this are following - 
*  The huge number of [supported devices](https://dbp.noobdev.io/supported_devices.html) (278).
* Does not require explicit support from ROMs

[Source](https://github.com/chenxiaolong/DualBootPatcher) and [XDA thread](https://forum.xda-developers.com/showthread.php?t=2447534)

**Comment by z3ntu** [comment]
@PabloCastellano Just want to correct you quickly

> It uses the Android build system (source)

EFIDroid does NOT use the Android build system, it uses similar syntax and targets and uses the repo tool from Google but the rest are his own Python scripts etc.

**Comment by eMPee584** [comment]
For the record: Safestrap (Droid 4 et al.) is a similar boot manager project..

**Comment by ollieparanoid** [comment]
@NotKit got safestrap working for the Droid 4 as [initfs hook](https://github.com/postmarketOS/pmbootstrap/pull/1039/files#diff-6cbc3d33c6e0d57f3518641b2c31e7b2).

**Comment by mirh** [comment]
Safestrap is more about working around a bootloader you can't control than anything. 

Then EFIDroid certainly looks the neater thingy imo, but it won't be available on all devices. 
And similarly phones may have all the weirdest combinations of /boot restrictions, from those with a partition too little to accommodate two kernels, to those unable to load kernels outside of it (at least in normal conditions without kexec). 

So imo, to avoid this kind of discussion x1000 there should be a list of "practices preferences", from those requiring the less burden/effort/changes/hassle/time-to-boot, to maybe [having a third kernel](https://forum.xda-developers.com/showthread.php?t=1880305) to manage boot selection, to overcomplicated one like safestrap or kexecboot.

**Comment by ollieparanoid** [comment]
Yeah the extra kernel approach with kexecboot for an universal bootloader that works from any other bootloader is also an interesting approach. That way we could provide a boot menu and offer choice between multiple kernels for all devices.
